### ISSUE TITLE
Should class declarations contain a binding for class name in scope for the class?

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#4161]
Should class declarations contain a binding for class name in the scope for the class?

`Klass` in example below has 2 bindings - 1st at top level, 2nd within the class body.

```js
class Klass {
    static tryWrite() {
        Klass = 123;
    }
}

const CopyOfKlass = Klass;

// Reassignment of `Klass` in top level scope succeeds
Klass = 123;
console.log(Klass); // Logs 123

// Reassignment of `Klass` within class fails.
// `TypeError: Assignment to constant variable at Klass.tryWrite`
CopyOfKlass.tryWrite();
```

The outer binding is reassignable, whereas the binding within the class is read-only.

**Comment by Dunqing** [comment]
In the spec. The `class` doesn't have a scope.

In the `TypeScript` container. Will create a container for `class`. But doesn't add a binding for class name to the `container`

**Comment by overlookmotel** [comment]
> In the spec. The class doesn't have a scope.

Is it possible for the spec to be wrong?? (or wrong in terms of what we need). I do think in this case there are 2 bindings. The above example demonstrates this.

**Comment by Dunqing** [comment]
> Is it possible for the spec to be wrong?? (or wrong in terms of what we need). 

Unlikely. But we are aligning with the TypeScript container now.

I do think in this case there are 2 bindings. The above example demonstrates this.

IMO, we don't need to add a binding to the class scope. The error in the above example is a runtime error that we don't need to care about

**Comment by rzvxa** [comment]
I think the class identifier much like the `this` keyword means something special in the context of class and that's why you can't reassign it

**Comment by overlookmotel** [comment]
Maybe the spec has some "special" way of thinking about this. But for our purposes, you have 2 variables with the same name which are independent of each other and one shadows the other - that sounds a lot like scopes to me!

**Comment by overlookmotel** [comment]
> The error in the above example is a runtime error that we don't need to care about

I think we will need to care about it in future. e.g. a transform which converts a `class` into a `function`. And minifier may be able to perform optimizations if it understands that one binding is read-only.

**Comment by rzvxa** [comment]
But it will cause the scopes to be wrong in some cases.

```js
class Klass {
  y() {
    class Foo {
      y() {
        console.log(Klass, Foo);
      }
    }
    return new Foo();
  }
}

new Klass().y().y();
```

What should `Klass` and `Foo` point to in the `console.log` line?

**Comment by overlookmotel** [comment]
> What should Klass and Foo point to in the `console.log` line?

Equivalent to:

```js
let KlassOuter = class KlassInner {
  y() {
    let FooOuter = class FooInner {
      y() {
        console.log(KlassInner, FooInner);
      }
    };
    return new FooOuter();
  }
};

new KlassOuter().y().y();
```

I did originally open this issue to pose a question, but now I've become convinced of the answer!

**Comment by rzvxa** [comment]
I'm not 100% sure if it is right or wrong but please investigate it more before doing anything. I feel TS people had a good reason to not create a scope here while they could do it easily. What are our other options?

```js
let KlassOuter = class KlassInner {
  y() {
    let FooOuter = class FooInner {
      y() {
        console.log(KlassInner, FooInner);
        console.log(KlassInner, FooInner);
        console.log(KlassInner, FooInner);
        console.log(KlassInner, FooInner);
      }
    };
    return new FooOuter();
  }
};

new KlassOuter().y().y();
```

In the example above how many times `KlassOuter` should be read? With this, we have to propagate shadowed class identifiers throughout our scope analysis.

**Comment by overlookmotel** [comment]
> I feel TS people had a good reason to not create a scope here while they could do it easily.

It could be they take a different approach, because they're trying to do something different. Their emphasis is on type checking, and the types of `KlassOuter` and `KlassInner` are the same, so no point differentiating them. But we are doing something different - e.g. minification, where whether bindings are read-only or not matters.

> In the example above how many times KlassOuter should be read?

Once I guess in `new KlassOuter()`.

> With this, we have to propagate shadowed class identifiers throughout our scope analysis.

How do you mean?

NB: Babel's transform of class private methods seems to support my view: [Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEAa0G8BQ1oQC5nQS2NAZgPaEAUAlIiqtAE4Cm6ArjQHbToAW2EAdAMRFCAbioBfKhiy5oA4uUrVaDZm1gjU48SAbRghAA4BPaAF44I-GYCMAJgDMIvSwiFtPEIQDmJPUZ6D5EyDdA0MyIWgAekj2GkY6JCA&debug=false&forceAllTransforms=false&modules=commonjs&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.24.7&externalPlugins=%40babel%2Fplugin-transform-private-methods%407.24.7&assumptions=%7B%7D) - `_X` in output represents the inner binding.

**Comment by rzvxa** [comment]
I was AFK for a few hours. But I'd like to investigate it when I find the time. But this [Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEAa0G8BQ1oQC5nQS2NAZgPaEAUAlIiqtNvtCegJ4AOApoXfALxfQBEsPhWTVR1AE6t0AV3EA7aOgAW2CADoAxEUIBuKqgC-SKkdQYsuaFuLlKoyTPlw9hpEZBTowQs0bResHrc0ACMAEwAzHrechCEHmoghADmJN6-atq2PLzpjGQ60AD0RYri0qxIQA&debug=false&forceAllTransforms=false&modules=commonjs&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.24.7&externalPlugins=%40babel%2Fplugin-transform-private-methods%407.24.7&assumptions=%7B%7D) suggests that it only uses alias variables for lowering static members to prevent the Class identifier to be shadowed.

**Comment by rzvxa** [comment]
[Here ](https://babeljs.io/repl#?browsers=ie%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEAa0G8BQ1oQC5nQS2NAZgPaEAUAlIiqtAE4Cm6ArjQHZwDcSVAvl7yA2jBCABwCe0ALwckAellwp0AIwAmAMydhLCIQEA6EIQDmJYeP1FSFSbaGixZdtHnR0NRnSRA&debug=false&forceAllTransforms=false&modules=commonjs&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.24.7&externalPlugins=%40babel%2Fplugin-transform-private-methods%407.24.7&assumptions=%7B%7D)is the same example without private members which doesn't create an `_X` anymore (while it still being used in public methods)

**Comment by overlookmotel** [comment]
Yes, it only creates the `_X` alias when the method is converted to a function outside of the class. That makes sense - when the method remains inside the class, it doesn't need to create an alias, because `X` inside the class refers to the inner binding.

My point is:

1. The private method example demonstrates that there are 2 independent bindings.
2. We are going to have to implement the class private methods transform, so we need this represented in our scope tree.

**Comment by rzvxa** [comment]
I'm thinking about it the other way around. Instead of creating a dummy scope in the semantics which potentially can cause issues in other use cases like linters, We may want to identify the private members and reassign them to a newly created symbol as we are doing the transformation.

I'm not saying having 2 bindings is wrong as the matter of fact it might be the right thing to do, but It may cause unnecessary issues in other parts so that's why I'm saying we should be more careful about the implications of doing it this way.

Even If it works but isn't up to spec it can cause difficulties for people when they are trying to use the semantics. Sadly I don't have the time to investigate it at the moment. So here are my 2 cents, IMHO it is one of those things that might pay off if we go with the spec.
You guys clearly have a much clearer picture of it. I'll let you decide on this one. I believe you'd go with the best approach⚓.

**Comment by overlookmotel** [comment]
Oh lord. I really didn't think this would be so controversial!

OK, I'm not great at reading the spec, but I've done my best. Here goes...

[15.7.15](https://tc39.es/ecma262/#sec-runtime-semantics-bindingclassdeclarationevaluation)

> ClassDeclaration : class BindingIdentifier ClassTail
> 1. Let className be the StringValue of BindingIdentifier.
> 2. Let value be ? ClassDefinitionEvaluation of ClassTail with arguments className and className.

[15.7.14](https://tc39.es/ecma262/#sec-runtime-semantics-classdefinitionevaluation)

> The syntax-directed operation ClassDefinitionEvaluation takes arguments classBinding (a String or undefined) and className
> ...
> 3. If classBinding is not undefined, then
> a. Perform ! classEnv.CreateImmutableBinding(classBinding, true).

[9.1.1](https://tc39.es/ecma262/#sec-the-environment-record-type-hierarchy)

> CreateImmutableBinding(N, S)
> Create a new but uninitialized immutable binding in an Environment Record. The String value N is the text of the bound name. If S is true then attempts to set it after it has been initialized will always throw an exception, regardless of the strict mode setting of operations that reference that binding.

Translating this into pseudo-code:

```rs
fn BindingClassDeclarationEvaluation(className: String) {
    ClassDefinitionEvaluation(className, className);
}

fn ClassDefinitionEvaluation(classBinding: String | undefined, className: String) {
    if classBinding != undefined {
        classEnv.CreateImmutableBinding(classBinding, true);
    }
}

fn CreateImmutableBinding(self: Env, N: String, S: bool) {
    // Create binding with name `N` in `self` (environment).
    // If `S` is true, assigning to the binding will throw.
}
```

So, following the logic, a class definition generates a binding in `classEnv`. And that binding is immutable and throws if you assign to it.

This cannot be referring to the outer binding because the outer binding is not immutable, and does not throw when you assign to it.

Like I said, I find it pretty hard reading the spec, so maybe I am interpreting it wrongly, but it does seem to me to suggest there *is* an inner binding for the class name in class declarations. If I have it wrong, please correct me.

**Comment by rzvxa** [comment]
Wow! Thanks for summarizing the spec so greatly. Talking about it in terms of spec gives me more confidence about this change.
Sorry, it took me a few hours to respond to your comment, Had to read the section `15.7` to be sure.

There was only one caveat that I could find with the limited time I've spent on it, A class heritage should only have access to the `LexicalEnvironment` (classEnv) and its outer `PrivateEnvironment` to prevent stuff like this in a nonstrict environment.

```js
class Klass extends ((x = String)) { y = "YES"; } 
```

If heritage has access to the private environment, `Klass` would have 2 fields `x` and `y`. But it is just a variable definition.

**Comment by overlookmotel** [comment]
The example you give is interesting.

```js
function outer() {
    class Klass extends ((this.x = "NO"), class {}) {}
}

const obj = {};
outer.call(obj);
console.log(obj.x); // Logs "NO"
```

i.e. `this` in the `extends` clause refers to `this` in the enclosing environment.

But...

```js
class Klass extends (Klass, class {}) {}
```

throws an error `ReferenceError: Cannot access 'Klass' before initialization`.

Conclusion: `Klass` is within scope of the `extends` clause (but in TDZ). But the `this` binding is bound within a further nested scope (the class body).

**Comment by Boshen** [comment]
Do we have a consensus on this issue?

**Comment by Boshen** [comment]
No consensus, close as stale.

We'll conform to the container concept in tsc.
